Fuzzing docs (#530)
software-mansion/protostar: website/docs/tutorials/07-testing/03-fuzzing/README.md

# Fuzzing

> :::warn

Reviewer: ```suggestion
:::warning
```
My bad. I always confuse `warn` and `warning` box.

> :::info
> This feature is actively developed and many new additions will land in future Protostar releases.
> :::

Reviewer: ```suggestion
:::warn
This feature is actively developed and many new additions will land in future Protostar releases.
:::
```

> Protostar tests can take parameters which makes such tests to be run in a _fuzzing mode_.

Reviewer: ```suggestion
Protostar tests can take parameters, which makes such tests to be run in a _fuzzing mode_.
```

In this mode, Protostar treats the test case parameters as a specification of the test case,
in the form of properties which it should satisfy,

> and tests that these properties hold in a large number of randomly generated input data.

Reviewer: ```suggestion
and tests that these properties hold in numerous randomly generated input data.
```

        assert_nn(new_res)
    end

    balance.write(new_res)
    return ()
end

@constructor
func constructor{
    syscall_ptr : felt*,
    pedersen_ptr : HashBuiltin*,
    range_check_ptr
}():
    balance.write(10000)
    return ()
end
```

### Unit testing

Let's first verify this function by writing a unit test in order to find the general property we are
testing for:

```cairo title="tests/test_main.cairo"
%lang starknet
from src.main import balance, withdraw
from starkware.cairo.common.cairo_builtins import HashBuiltin

@external
func test_withdraw{
    syscall_ptr : felt*,
    pedersen_ptr : HashBuiltin*,
    range_check_ptr
}():
    alloc_locals
    let (pre_balance_ref) = balance.read()
    local pre_balance = pre_balance_ref

    let amount = 1

    withdraw(amount)

    let (post_balance) = balance.read()
    assert post_balance = pre_balance - amount

    return ()
end
```

So far, so good. Running the test, we see it passes:

```text title="$ protostar test"
[PASS] tests/test_main.cairo test_withdraw (steps=129)
       range_check_builtin=1

11:28:43 [INFO] Test suites: 1 passed, 1 total
11:28:43 [INFO] Tests:       1 passed, 1 total
```

### Generalizing the test

> This unit test does test that we can withdraw "some" amount from our safe.

Reviewer 2: ```suggestion
This unit test performs cheks if we can withdraw "some" amount from our safe.
```

> However, can we be sure that it works for all amounts, not just 1?

Reviewer 2: ```suggestion
However, can we be sure that it works for all amounts, not just this particular one?
```
it may suggests you have `1` as number in mind

[type] TRANSACTION_FAILED
[code] 39
[messages]:
— Cannot withdraw more than stored.
[details]:
<REDACTED>/starkware/cairo/common/math.cairo:40:5: Error at pc=0:0:
Got an exception while executing a hint.
    %{
    ^^
Cairo traceback (most recent call last):
tests/test_main.cairo:22:6: (pc=0:141)
func test_withdraw{syscall_ptr : felt*, range_check_ptr, pedersen_ptr : HashBuiltin*}(
     ^***********^
tests/test_main.cairo:36:5: (pc=0:125)
    withdraw(amount)
    ^**************^
Error message: Cannot withdraw more than stored.
<REDACTED>/src/main.cairo:23:9: (pc=0:63)
        assert_nn(new_res)
        ^****************^

Traceback (most recent call last):
  File "<REDACTED>/starkware/cairo/common/math.cairo", line 43, in <module>
    assert 0 <= ids.a % PRIME < range_check_builtin.bound, f'a = {ids.a} is out of range.'
AssertionError: a = 3618502788666131213697322783095070105282824848410658236509717448704103819025 is out of range.
[falsifying example]:
amount = 340282366920938463463374607431768211456

11:41:48 [INFO] Test suites: 1 failed, 1 total
11:41:48 [INFO] Tests:       1 failed, 1 total
11:41:48 [INFO] Seed:        4258368192
```

### Fixing the bug

> The test fails because if `amount` value is high enough, because it is a `felt`, when it is
> interpreted as a signed number in range comparisons, it overflows.

Reviewer 2: This sentence is a bit weird. What do you think about:

```suggestion
The test fails because `amount` has `felt` type so its value can be negative. If smallest possible `felt` value is substracted from `balance` it causes `felt` overflow.
```

func withdraw{syscall_ptr : felt*, pedersen_ptr : HashBuiltin*, range_check_ptr}(amount : felt):
    with_attr error_message("Amount must be positive."):
        assert_nn(amount)
    end
    
    # ...
end
```

```cairo title="tests/test_main.cairo"
@external
func test_withdraw{
    syscall_ptr : felt*,
    pedersen_ptr : HashBuiltin*,
    range_check_ptr
}(
    amount : felt
):
    # ...
    
    %{
        if not (0 <= ids.amount and ids.amount % PRIME < range_check_builtin.bound):
            expect_revert(error_message="Amount must be positive.")
        elif ids.amount > ids.pre_balance:
            expect_revert(error_message="Cannot withdraw more than stored.")
    %}
    withdraw(amount)
    
    # ...
end
```

And now, the test passes.
We can also observe the variance of resources usage, caused by the `if amount == 0:` branch in code.

```text title="$ protostar test"
[PASS] tests/test_main.cairo test_withdraw (steps=μ: 127, Md: 137, min: 84, max: 137)
       range_check_builtin=μ: 1.81, Md: 2, min: 1, max: 2

11:55:18 [INFO] Test suites: 1 passed, 1 total
11:55:18 [INFO] Tests:       1 passed, 1 total
11:55:18 [INFO] Seed:        3741774783
```

## Interpreting results

> In fuzzing mode, the test is executed many times, hence test summaries are slightly more extensive:

Reviewer 2: ```suggestion
In fuzzing mode, the test is executed many times, hence test summaries are extended:
```
